knative/deployer.go

```diff
@@ -69,39 +69,38 @@ func (d *Deployer) Deploy(ctx context.Context, f bosonFunc.Function) (err error)
 			err, _ = client.WaitForService(f.Name, DefaultWaitingTimeout, wait.NoopMessageCallback())
 			if err != nil {
 				err = fmt.Errorf("knative deployer failed to wait for the service to become ready: %v", err)
-				return err
+				return "", err
 			}
 
 			route, err := client.GetRoute(f.Name)
 			if err != nil {
 				err = fmt.Errorf("knative deployer failed to get the route: %v", err)
-				return err
+				return "", err
 			}
 
 			fmt.Println("Function deployed at URL: " + route.Status.URL.String())
+			return fmt.Sprintf("Function deployed at URL: %v", route.Status.URL.String()), nil
 
 		} else {
 			err = fmt.Errorf("knative deployer failed to get the service: %v", err)
-			return err
+			return "", err
 		}
 	} else {
 		// Update the existing Service
 		err = client.UpdateServiceWithRetry(f.Name, updateService(f.ImageWithDigest(), f.Env, f.Annotations), 3)
 		if err != nil {
 			err = fmt.Errorf("knative deployer failed to update the service: %v", err)
-			return err
+			return "", err
 		}
 
 		route, err := client.GetRoute(f.Name)
 		if err != nil {
 			err = fmt.Errorf("knative deployer failed to get the route: %v", err)
-			return err
+			return "", err
 		}
 
-		fmt.Println("Function updated at URL: " + route.Status.URL.String())
+		return fmt.Sprintf("Function updated at URL: %v", route.Status.URL.String()), nil
```

Comment:
This return value should probably be the route data (url) only, and the extra natural language English prefix added in the CLI command `cmd/deploy.go`, because that is essentially the UX, whereas here is more for data / client integration.
```suggestion
		return route.Status.URL.String(), nil
```

Reply:
This is a good point. Going this route, we would probably want to remove this as well: https://github.com/boson-project/func/pull/322/files#diff-3bc17fd6241d546478336fa19a33f7e7f4c232421ce4e8511512c00fa617860dR52-R54. One side effect of this change however is that `Deployer` is the only part that knows if the function service was created or updated. We would need to either expose this property somehow, or forego the language.